### PR TITLE
Cluster connectivity check

### PR DESCRIPTION
**Pull request by hanefi**
DESCRIPTION: Introduces `citus_check_cluster_node_health` UDF to check cluster connectivity

This PR aims to create an easy to use API for coordinating connectivity checks between all node pairs in the cluster.

# Usage

`citus_check_cluster_node_health()` takes no parameters and checks the connectivity between pairs of nodes sequentially.

e.g.
```
 from_nodename | from_nodeport | to_nodename | to_nodeport | result
--------------------------------------------------------------------
 localhost     |          9700 | localhost   |        9700 | t
 localhost     |          9700 | localhost   |        9701 | t
 localhost     |          9701 | localhost   |        9700 | t
 localhost     |          9701 | localhost   |        9701 | t
(4 rows)
```

## Return value

This UDF returns a set of records indicating the source and target nodes and success status.

There are 3 possible success status:
1. `true` : connection attempt from source to target succeeded.
2. `false`: connection attempt from source to target failed.
3. `NULL` : connection attempt from the current node to source node failed.

In the case of a `NULL` or `false` success value, we log `WARNING` messages for the underlying connection issues.

## Suggested Usage

We suggest operators to use the following command to check connectivity in the cluster.

```sql
SELECT bool_and(coalesce(result, false)) FROM citus_check_cluster_node_health();
 bool_and
---------------------------------------------------------------------
 t
(1 row)

```

# The algorithm

```
for sourceNode in workerNodeList:
	c = connectToNode(sourceNode)
	for targetNode in workerNodeList:
		result = c.execute(
			"SELECT citus_check_connection_to_node(targetNode.name,
												   targetNode.port")
		emit sourceNode.name,
			 sourceNode.port,
			 targetNode.name,
			 targetNode.port,
			 result
```

Related: #5487
Fixes: #5290

### CODE REVIEW

File: src/test/regress/expected/multi_citus_tools.out
Comment (hanefi) [inline src/test/regress/expected/multi_citus_tools.out:644]: This test has a broken result when run on the test suite.

On a local cluster, I see the expected results:
```
hanefi=# SELECT * FROM citus_check_cluster_node_health();
LOG:  statement: SELECT * FROM citus_check_cluster_node_health();
NOTICE:  issuing SELECT citus_check_connection_to_node('localhost', 9701)
DETAIL:  on server hanefi@localhost:9701 connectionId: 1
DEBUG:  issuing SELECT 1
DETAIL:  on server hanefi@localhost:9701 connectionId: 9
DETAIL:  from localhost:9701
NOTICE:  issuing SELECT citus_check_connection_to_node('localhost', 9702)
DETAIL:  on server hanefi@localhost:9701 connectionId: 1
DEBUG:  issuing SELECT 1
DETAIL:  on server hanefi@localhost:9702 connectionId: 10
DETAIL:  from localhost:9701
NOTICE:  issuing SELECT citus_check_connection_to_node('localhost', 9701)
DETAIL:  on server hanefi@localhost:9702 connectionId: 2
DEBUG:  issuing SELECT 1
DETAIL:  on server hanefi@localhost:9701 connectionId: 9
DETAIL:  from localhost:9702
NOTICE:  issuing SELECT citus_check_connection_to_node('localhost', 9702)
DETAIL:  on server hanefi@localhost:9702 connectionId: 2
DEBUG:  issuing SELECT 1
DETAIL:  on server hanefi@localhost:9702 connectionId: 10
DETAIL:  from localhost:9702
 from_nodename | from_nodeport | to_nodename | to_nodeport | success
---------------+---------------+-------------+-------------+---------
 localhost     |          9701 | localhost   |        9701 | t
 localhost     |          9701 | localhost   |        9702 | t
 localhost     |          9702 | localhost   |        9701 | t
 localhost     |          9702 | localhost   |        9702 | t
```

File: src/test/regress/expected/multi_citus_tools.out
Comment (onderkalaci) [inline src/test/regress/expected/multi_citus_tools.out:632]: seems like a good candidate to add to `arbitrary-configs` like `SELECT bool_and(success) citus_check_cluster_node_health()`

File: src/backend/distributed/operations/citus_tools.c
Comment (onderkalaci) [inline src/backend/distributed/operations/citus_tools.c:159]: minor: given the multiple loops, lets give more descriptive names: connectionToSourceNode, sourceWorkerNode, sourceWorkerNodeList, targetWorkerNodeList, connectivityCheckFromSourceToTargetCommand or such.

File: src/test/regress/expected/multi_citus_tools.out
Comment (hanefi) [reply src/test/regress/expected/multi_citus_tools.out:632]: This is a really good advice. I still have not figured out how to make my tests pass on all configs. Working on it

File: src/test/regress/expected/multi_citus_tools.out
Comment (hanefi) [reply src/test/regress/expected/multi_citus_tools.out]: It turned out that I was missing some calls to `FinishConnectionEstablishment`. Fixed that in 658cb77

File: src/backend/distributed/operations/citus_tools.c
Comment (hanefi) [reply src/backend/distributed/operations/citus_tools.c:173]: We already do. I iterate on the same `workerNodeList` that contains all active primary nodes, and use `sourceWorkerNode` and `targetWorkerNode` variables.

File: src/backend/distributed/operations/citus_tools.c
Comment (hanefi) [reply src/backend/distributed/operations/citus_tools.c]: I updated the variable names, and added a summary of the algorithm in the comments to make it easier to follow.

File: src/backend/distributed/operations/citus_tools.c
Comment (hanefi) [reply src/backend/distributed/operations/citus_tools.c]: Thanks for this explanation. I wonder if it makes sense to print a better warning message if we haven't finished connection establishment.

In theory that should never happen. However, it can also help us find bugs and make development easier. What do you think @onderkalaci ?

File: src/backend/distributed/operations/citus_tools.c
Comment (onderkalaci) [inline src/backend/distributed/operations/citus_tools.c:178]: There is one topic that I have been thinking about this. This gives the connection pairs from the coordinator's perspective of the cluster (e.g., using coordinator's metadatta). In reality, it should have been per-node's perspective of the cluster.

But, given that all nodes should have the same metadata with the current architecture, we should be fine.

Do you agree?

File: src/backend/distributed/operations/citus_tools.c
Comment (onderkalaci) [inline src/backend/distributed/operations/citus_tools.c:39]: Minor: Should we move these new UDFs into own source code file? I don't think `master_run_on_worker` is very much related to what we add here. Probably best to do with a follow-up PR.

File: src/backend/distributed/operations/citus_tools.c
Comment (onderkalaci) [reply src/backend/distributed/operations/citus_tools.c:173]: I agree that there is no correctness issue. But, don't you think that there is a readability issue here?

I think you have a good understanding of how `foreach_ptr` works, but for readers without this information, it is very confusing to iterate over the same list.

It at least needs a comment at the top of the second loop, saying that it is safe to iterate on the same list as long as we use different variable while iterating.

File: src/backend/distributed/operations/citus_tools.c
Comment (onderkalaci) [reply src/backend/distributed/operations/citus_tools.c]: I think a good improvement could be to make `StartNodeConnection ` a static function at least.

And, yes, we could consider add some sanity checks if you use a connection that is not successfully established yet.

But, probably in a separate PR

File: src/backend/distributed/operations/citus_tools.c
Comment (onderkalaci) [inline src/backend/distributed/operations/citus_tools.c:234]: is it possible to either (a) undo these unrelated changes (b) have them in a separate commit? Mostly asking for a cleaner git history purposes.

File: src/test/regress/expected/failure_connection_establishment.out
Comment (onderkalaci) [inline src/test/regress/expected/failure_connection_establishment.out:325]: Hmm, this is actually an unexpected/misleading user experience. The connectivity between the coordinator and 9060 is broken. But, we show `9060` -> `9060` and `9060`-> `57637` connectivity checks as broken. This could lead to very hard times while debugging actual connectivity issues.

We should probably check the return code of `ExecuteOptionalRemoteCommand`. If the return value is NOT `RESPONSE_OKAY`, we should report that connection is also broken. If the local host that we are running is already in PgDistNode, we can actually skip reporting these connectivity checks, because that failure is already going to be reported.

What we can do is something like:
```C

    if response != RESPONSE_OKAY:
            /* we'll going to report this failure anyway, so for these nodes just return NULL indicating that we didn't successfully connect to the source node */
           success = NULL;
          if MyLocalGroupIdInPgDistNode:
                /* we don't need anything special, because this connection pair would be reported anyway*/
          else: 
                add a new row to the output from: LocalHostName GUC / PG knows the current servers, we already know to the node that we fail to connect, false
```

What are your thoughts on this?

Comment (onderkalaci) [reply src/test/regress/expected/failure_connection_establishment.out:325]: Another discussion on this can be found here: https://github.com/citusdata/citus/issues/5290#issuecomment-942122862

File: src/test/regress/expected/multi_citus_tools.out
Comment (hanefi) [reply src/test/regress/expected/multi_citus_tools.out:632]: Do you think that it is enough to see the result of the following?

```sql
SELECT bool_and(success) FROM citus_check_cluster_node_health();
```

I did not want to add complex checks as we have many configs now.

File: src/backend/distributed/operations/citus_tools.c
Comment (hanefi) [reply src/backend/distributed/operations/citus_tools.c:178]: I agree. The coordinator should decide on the node pairs. If one of the worker nodes is missing some `pg_dist_node` records, we may end up skipping some of the checks, and potentially give a false success message when we run:

	SELECT bool_and(success) FROM citus_check_cluster_node_health();

File: src/test/regress/expected/multi_citus_tools.out
Comment (onderkalaci) [reply src/test/regress/expected/multi_citus_tools.out:632]: Yes, that is what I had actually.

File: src/test/regress/expected/multi_citus_tools.out
Comment (onderkalaci) [inline src/test/regress/expected/multi_citus_tools.out:647]: Cool test, would be good to have NULL values in the hostname/port for extreme testing.

File: src/backend/distributed/operations/citus_tools.c
Comment (hanefi) [reply src/backend/distributed/operations/citus_tools.c:39]: I should do that after merging this PR. This is a really good advice.

The connectivity checks used to depend on the static functions here. Now it does not really make sense to have them here.

File: src/backend/distributed/operations/citus_tools.c
Comment (hanefi) [reply src/backend/distributed/operations/citus_tools.c]: I removed these changes. I do not think that I should update these lines at all.

It should be fine now.

File: src/test/regress/expected/failure_connection_establishment.out
Comment (hanefi) [inline src/test/regress/expected/failure_connection_establishment.out:347]: @onderkalaci it is no longer feasible to use `bool_and()` to create a summary of the connectivity check. Here is an example, why that is necessary.

Please let me know if you think we should not have these lines in failure tests.

File: src/test/regress/expected/multi_citus_tools.out
Comment (hanefi) [reply src/test/regress/expected/multi_citus_tools.out:647]: Do you think that I should test for NULL values, even though `nodename` and `nodeport` are not NULLABLE?

Can you think of an easy way to check for connections to nodes with NULL nodenames, without adding new entries to `pg_dist_node`?
```
                                  Table "pg_catalog.pg_dist_node"
      Column      |   Type   | Collation | Nullable |                   Default
------------------+----------+-----------+----------+----------------------------------------------
 nodeid           | integer  |           | not null | nextval('pg_dist_node_nodeid_seq'::regclass)
 groupid          | integer  |           | not null | nextval('pg_dist_groupid_seq'::regclass)
 nodename         | text     |           | not null |
 nodeport         | integer  |           | not null | 5432
 noderack         | text     |           | not null | 'default'::text
 hasmetadata      | boolean  |           | not null | false
 isactive         | boolean  |           | not null | true
 noderole         | noderole |           | not null | 'primary'::noderole
 nodecluster      | name     |           | not null | 'default'::name
 metadatasynced   | boolean  |           |          | false
 shouldhaveshards | boolean  |           | not null | true
 ```

Comment (hanefi) [reply src/test/regress/expected/multi_citus_tools.out:647]: If we have any NULL values, the defaults are used. Only the `nodename` does not have a default, and should always be supplied.

I could easily drop `5432` nodeport in this example to partially cover what you proposed in this comment.

File: src/backend/distributed/operations/citus_tools.c
Comment (onderkalaci) [inline src/backend/distributed/operations/citus_tools.c:119]: minor: I think `tuplestore_donestoring` is not needed `#define tuplestore_donestoring(state) ((void) 0)`

File: src/backend/distributed/operations/citus_tools.c
Comment (onderkalaci) [inline src/backend/distributed/operations/citus_tools.c:157]: What about read-replicas? It seems useful that this UDF works there as well.

I think we can probably use `ActiveReadableNodeList` instead of `ActivePrimaryNodeList`. But, we should also add one more test to somewhere like `multi_follower_select_statements.sql`

File: src/test/regress/expected/multi_citus_tools.out
Comment (onderkalaci) [reply src/test/regress/expected/multi_citus_tools.out:647]: Given that we are in a TX, we can drop NOT NULL constraint:
```SQL
BEGIN;

alter table pg_dist_node alter COLUMN nodename DROP NOT NULL ;

 insert into pg_dist_node(nodename,nodeport) VALUES (NULL, 5432);
INSERT 0 1
Time: 13.709 ms
select * from citus_check_cluster_node_health(); 
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.
Time: 2.603 ms
```
which seems to be crashing. So, we should be fixing it and adding the test then :)

Comment (onderkalaci) [reply src/test/regress/expected/multi_citus_tools.out:647]: The fix could be just hard errors for this unexpected case.

File: src/test/regress/create_schedule
Comment (onderkalaci) [inline src/test/regress/create_schedule:6]: if we put this also to `sql_schedule`, it'd also run from the workers, which is a good coverage. 

If there are non-mx workers, this test might fail. But, as far as I remember all the tests run with MX enabled. So, can you also try adding there? If anything fails, I think running `citus_check_cluster_node_health` from the workers on one of the mx tests would be good sanity check

File: src/test/regress/expected/connectivity_checks.out
Comment (onderkalaci) [inline src/test/regress/expected/connectivity_checks.out:1]: I wonder if our *official* suggestion for querying the cluster health should be:
```SQL
-- returns sourceHost:sourcePort targetHost:targetPort success
-- success -> true -> connection attempt from source to target succeeded
-- success -> false -> connection attempt from source to target failed
-- success -> NULL -> connection attempt from the current node to target failed 
select bool_and(COALESCE(success, false)) from citus_check_cluster_node_health();
```

Whenever this query return false, there is a connectivity issue, check in detail.

**Review by onderkalaci — COMMENTED** [review]
Mostly looks really hood, probably we are in the final iteration of this review

File: src/test/regress/expected/connectivity_checks.out
Comment (onderkalaci) [reply src/test/regress/expected/connectivity_checks.out:1]: also, maybe we should rename `success ` -> `result` ?

File: src/backend/distributed/operations/citus_tools.c
Comment (hanefi) [reply src/backend/distributed/operations/citus_tools.c:119]: Apparently, PG do not need this call since 7.4 release.

What is really strange is that Tom Lane committed a change last year using this NOOP call https://github.com/postgres/postgres/commit/dd04e958c8b03c0f0512497651678c7816af3198

I will remove the call here, and maybe remove 20 other occurrences spread across 16 files on a separate PR.

File: src/backend/distributed/operations/citus_tools.c
Comment (onderkalaci) [reply src/backend/distributed/operations/citus_tools.c:157]: Do you know how to setup Citus-ha with read-replicas? Once you change this, it'd be great to test. Let me know if you need any help testing this

File: src/test/regress/create_schedule
Comment (hanefi) [reply src/test/regress/create_schedule:6]: There are many inaccessible worker nodes with different cluster names in MX tests.

```
SELECT * FROM pg_dist_node;
 nodeid | groupid | nodename  | nodeport | noderack | hasmetadata | isactive |  noderole   |                           nodecluster                           | metadatasynced | shouldhaveshards
---------------------------------------------------------------------
     16 |      14 | localhost |    57637 | default  | t           | t        | primary     | default                                                         | t              | t
     25 |      14 | localhost |     8888 | default  | f           | t        | secondary   | olap                                                            | f              | t
     30 |      18 | localhost |    57638 | default  | t           | t        | primary     | default                                                         | t              | t
     26 |      14 | localhost |     8887 | default  | f           | t        | secondary   | thisisasixtyfourcharacterstringrepeatedfourtimestomake256chars. | f              | t
     27 |      14 | localhost |     9995 | default  | f           | t        | secondary   | default                                                         | f              | t
     28 |      16 | localhost |     9994 | default  | f           | t        | secondary   | default                                                         | f              | t
     29 |      14 | localhost |     9992 | default  | f           | t        | secondary   | second-cluster                                                  | f              | t
     19 |      14 | localhost |     9998 | default  | f           | t        | secondary   | default                                                         | f              | t
     20 |      14 | localhost |     9997 | default  | f           | t        | unavailable | default                                                         | f              | t
     21 |      16 | localhost |     9996 | default  | f           | f        | secondary   | default                                                         | f              | t
(10 rows)
```

I ended up comparing the distinct counts of connectivity check results in the coordinator and the mx worker node.

```
SELECT result, count(*) FROM citus_check_cluster_node_health() GROUP BY result ORDER BY result;
 result | count
---------------------------------------------------------------------
 f      |    16
 t      |     4
        |    80
(3 rows)
```

File: src/backend/distributed/operations/citus_tools.c
Comment (hanefi) [reply src/backend/distributed/operations/citus_tools.c]: I ended up reading directly from the catalogs as our cache does not include workers from other clusters.

AFAIK we have some users that set the same `nodecluster` values of a set of read replicas in a cluster. Reading from catalog seemed to be the best way to cover connectivity checks to those nodes as well.

**Comment by hanefi** [comment]
@onderkalaci I guess I addressed all your comments.

There is only one comment that I intentionally do not address here. I do not think that I should add warning messages for null node names, as the `WorkerNode` cache breaks, and backend crashes. Also we no longer use the caches here.


a simple test:

```
hanefi=# BEGIN;
BEGIN
hanefi=*# alter table pg_dist_node alter COLUMN nodename DROP NOT NULL ;
ALTER TABLE
hanefi=*#  insert into pg_dist_node(nodename,nodeport) VALUES (NULL, 5432);
INSERT 0 1
hanefi=*# CREATE TABLE t(id int);
server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.
!?>
```

Assertion failure details:
```c

/*
 * TupleToWorkerNode takes in a heap tuple from pg_dist_node, and
 * converts this tuple to an equivalent struct in memory. The function assumes
 * the caller already has locks on the tuple, and doesn't perform any locking.
 */
static WorkerNode *
TupleToWorkerNode(TupleDesc tupleDescriptor, HeapTuple heapTuple)
{
	Datum datumArray[Natts_pg_dist_node];
	bool isNullArray[Natts_pg_dist_node];

	Assert(!HeapTupleHasNulls(heapTuple));
```

call stack:

```
libc.so.6!__GI_raise(int sig) (\build\glibc-vjB4T1\glibc-2.28\sysdeps\unix\sysv\linux\raise.c:50)
libc.so.6!__GI_abort() (\build\glibc-vjB4T1\glibc-2.28\stdlib\abort.c:79)
ExceptionalCondition(const char * conditionName, const char * errorType, const char * fileName, int lineNumber) (\home\hanefi\.pgenv\src\postgresql-14.0\src\backend\utils\error\assert.c:69)
citus.so!TupleToWorkerNode(TupleDesc tupleDescriptor, HeapTuple heapTuple) (\home\hanefi\code\citus\src\backend\distributed\metadata\node_metadata.c:2189)
citus.so!ReadDistNode(_Bool includeNodesFromOtherClusters) (\home\hanefi\code\citus\src\backend\distributed\metadata\node_metadata.c:1327)
citus.so!InitializeWorkerNodeCache() (\home\hanefi\code\citus\src\backend\distributed\metadata\metadata_cache.c:3437)
citus.so!PrepareWorkerNodeCache() (\home\hanefi\code\citus\src\backend\distributed\metadata\metadata_cache.c:3398)
citus.so!GetWorkerNodeHash() (\home\hanefi\code\citus\src\backend\distributed\metadata\metadata_cache.c:3369)
citus.so!PrimaryNodeForGroup(int32 groupId, _Bool * groupContainsNodes) (\home\hanefi\code\citus\src\backend\distributed\metadata\node_metadata.c:832)
citus.so!CoordinatorAddedAsWorkerNode() (\home\hanefi\code\citus\src\backend\distributed\operations\worker_node_manager.c:426)
citus.so!ShouldEnableLocalReferenceForeignKeys() (\home\hanefi\code\citus\src\backend\distributed\commands\table.c:316)
citus.so!PostprocessCreateTableStmtForeignKeys(CreateStmt * createStatement) (\home\hanefi\code\citus\src\backend\distributed\commands\table.c:257)
citus.so!PostprocessCreateTableStmt(CreateStmt * createStatement, const char * queryString) (\home\hanefi\code\citus\src\backend\distributed\commands\table.c:217)
citus.so!ProcessUtilityInternal(PlannedStmt * pstmt, const char * queryString, ProcessUtilityContext context, ParamListInfo params, struct QueryEnvironment * queryEnv, DestReceiver * dest, QueryCompletion * completionTag) (\home\hanefi\code\citus\src\backend\distributed\commands\utility_hook.c:659)
citus.so!multi_ProcessUtility(PlannedStmt * pstmt, const char * queryString, _Bool readOnlyTree, ProcessUtilityContext context, ParamListInfo params, struct QueryEnvironment * queryEnv, DestReceiver * dest, QueryCompletion * completionTag) (\home\hanefi\code\citus\src\backend\distributed\commands\utility_hook.c:266)
citus.so!ColumnarProcessUtility(PlannedStmt * pstmt, const char * queryString, _Bool readOnlyTree, ProcessUtilityContext context, ParamListInfo params, struct QueryEnvironment * queryEnv, DestReceiver * dest, QueryCompletion * completionTag) (\home\hanefi\code\citus\src\backend\columnar\columnar_tableam.c:2116)
pg_stat_statements.so!pgss_ProcessUtility(PlannedStmt * pstmt, const char * queryString, _Bool readOnlyTree, ProcessUtilityContext context, ParamListInfo params, QueryEnvironment * queryEnv, DestReceiver * dest, QueryCompletion * qc) (\home\hanefi\.pgenv\src\postgresql-14.0\contrib\pg_stat_statements\pg_stat_statements.c:1131)
ProcessUtility(PlannedStmt * pstmt, const char * queryString, _Bool readOnlyTree, ProcessUtilityContext context, ParamListInfo params, QueryEnvironment * queryEnv, DestReceiver * dest, QueryCompletion * qc) (\home\hanefi\.pgenv\src\postgresql-14.0\src\backend\tcop\utility.c:523)
PortalRunUtility(Portal portal, PlannedStmt * pstmt, _Bool isTopLevel, _Bool setHoldSnapshot, DestReceiver * dest, QueryCompletion * qc) (\home\hanefi\.pgenv\src\postgresql-14.0\src\backend\tcop\pquery.c:1147)
PortalRunMulti(Portal portal, _Bool isTopLevel, _Bool setHoldSnapshot, DestReceiver * dest, DestReceiver * altdest, QueryCompletion * qc) (\home\hanefi\.pgenv\src\postgresql-14.0\src\backend\tcop\pquery.c:1304)
PortalRun(Portal portal, long count, _Bool isTopLevel, _Bool run_once, DestReceiver * dest, DestReceiver * altdest, QueryCompletion * qc) (\home\hanefi\.pgenv\src\postgresql-14.0\src\backend\tcop\pquery.c:786)
exec_simple_query(const char * query_string) (\home\hanefi\.pgenv\src\postgresql-14.0\src\backend\tcop\postgres.c:1214)
PostgresMain(int argc, char ** argv, const char * dbname, const char * username) (\home\hanefi\.pgenv\src\postgresql-14.0\src\backend\tcop\postgres.c:4486)
BackendRun(Port * port) (\home\hanefi\.pgenv\src\postgresql-14.0\src\backend\postmaster\postmaster.c:4506)
BackendStartup(Port * port) (\home\hanefi\.pgenv\src\postgresql-14.0\src\backend\postmaster\postmaster.c:4228)
ServerLoop() (\home\hanefi\.pgenv\src\postgresql-14.0\src\backend\postmaster\postmaster.c:1745)
PostmasterMain(int argc, char ** argv) (\home\hanefi\.pgenv\src\postgresql-14.0\src\backend\postmaster\postmaster.c:1417)
main(int argc, char ** argv) (\home\hanefi\.pgenv\src\postgresql-14.0\src\backend\main\main.c:209)
```

File: src/backend/distributed/operations/citus_tools.c
Comment (onderkalaci) [reply src/backend/distributed/operations/citus_tools.c]: > AFAIK we have some users that set the same nodecluster values of a set of read replicas in a cluste

Short answer: We should switch back to `ActiveReadableNodeList ` :) 

Long answer:

I really don't think this is a good idea & even exists. 

Even if you have multiple read-replicas, you can only query one of the clusters at a time. And, typically you cannot change this because it is `PGC_SU_BACKEND`. So, if you want to query the other replica, you should do:
```
ALTER SYSTEM SET citus.cluster_name TO ...
SELECT pg_reload_conf();
....
-- reconnect via psql
SELECT query....
```

Also, there is no mechanism in the code that can handle multiple read-replicas at a time. `ReadNodeList(true)` is only used once in the code, when metadata is syned. Also, `ClusterName` expects a single cluster name anyway.

Also, see how the planner/executor assigns task placements. It is always relied on as `ActiveReadableNodeList `. And, our aim with this UDF is to signal that query executions can be done fine when returns true.

**Comment by onderkalaci** [comment]
> Assertion failure details:

Alright, then I'm fine with not acting on this assertion failure given that it requires changing the metadata.

File: src/test/regress/expected/multi_follower_select_statements.out
Comment (onderkalaci) [inline src/test/regress/expected/multi_follower_select_statements.out:163]: Ops, this looks very unusual output? Shouldn't we have 4 rows?  Is it leftover from the previous nodeList in the code? It'd be nice to do `SELECT * FROM citus_check_cluster_node_health()`

Comment (hanefi) [reply src/test/regress/expected/multi_follower_select_statements.out:163]: This is a follower test, and in this test setup we have followers in the cluster.

2 primary workers + 2 followers = 4
4 x 4 = 16

Let me use your suggested query and update the comments.

File: src/test/regress/multi_mx_schedule
Comment (onderkalaci) [inline src/test/regress/multi_mx_schedule:28]: would have been great to show that we can access to the coordinator if the coordinator is in the metadata.
Would you mind inlining the test file `multi_mx_connectivity` into `multi_mx_add_coordinator` so that we can show it?

**Review by onderkalaci — APPROVED** [review]
With a minor comment, looks pretty good. The test coverage is great.

It might be good to see the execution times with a large cluster, though as we'll do such large benchmarks later in the release, I save this idea there.

**Pull request merged**
